**1. DESCRIBE THE DATASET**

**2. IMPORT THE DATASET TO A PANDAS DATAFRAME**

In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")

**2.A Describe the dataset using the functions available in pandas**

*note for self:* describe what the functions do and why we chose these ones.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
df.shape

(891, 12)

In [5]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

**2.B What data types do you find?**

As seen in cell 3, we have several different types of data. Most are intagers (int64) which are whole numbers. There are also floats (float64) which are numbers with possibility of decimals. Lastly we have strings (object) which are words and letters. 

**2.C Are all records complete or are some data missing?**

As we can see in the "Non-null" column of the info table, there are missing data in Age, Cabin and Embarked. Since cabin have so many missing, and we can't recreate the data in any way, for the predictions for Kaggle, will mostly work on filling in the ages and embarked in new columns called ages_filled and embarked_filled, to use for later calculations. 

**3. FIND INFORMATION ABOUT THE PEOPLE IN THE TRAGEDY**

*Note to self:* add descriptions later

In [6]:
df.value_counts("Survived")

Survived
0    549
1    342
dtype: int64

In [7]:
df.Age.describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

In [8]:
df.groupby("Pclass")["Age"].mean()

Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64

**4. LIST PASSENGERS SORTED BY FAMILY NAME**

In [9]:
df["Last_name"] = df["Name"].str.split(",").str[0]

In [10]:
df.sort_values("Last_name")

PassengerId  Survived  Pclass                                   Name  \
845          846         0       3                    Abbing, Mr. Anthony   
279          280         1       3       Abbott, Mrs. Stanton (Rosa Hunt)   
746          747         0       3            Abbott, Mr. Rossmore Edward   
874          875         1       2  Abelson, Mrs. Samuel (Hannah Wizosky)   
308          309         0       2                    Abelson, Mr. Samuel   
..           ...       ...     ...                                    ...   
286          287         1       3                de Mulder, Mr. Theodore   
282          283         0       3              de Pelsmaeker, Mr. Alfons   
361          362         0       2              del Carlo, Mr. Sebastiano   
153          154         0       3        van Billiard, Mr. Austin Blyler   
868          869         0       3            van Melkebeke, Mr. Philemon   

        Sex   Age  SibSp  Parch         Ticket     Fare Cabin Embarked  \
845    male  42.0      0      0      C.A. 5547   7.5500   NaN        S   
279  female  35.0      1      1      C.A. 2673  20.2500   NaN        S   
746    male  16.0      1      1      C.A. 2673  20.2500   NaN        S   
874  female  28.0      1      0      P/PP 3381  24.0000   NaN        C   
308    male  30.0      1      0      P/PP 3381  24.0000   NaN        C   
..      ...   ...    ...    ...            ...      ...   ...      ...   
286    male  30.0      0      0         345774   9.5000   NaN        S   
282    male  16.0      0      0         345778   9.5000   NaN        S   
361    male  29.0      1      0  SC/PARIS 2167  27.7208   NaN        C   
153    male  40.5      0      2       A/5. 851  14.5000   NaN        S   
868    male   NaN      0      0         345777   9.5000   NaN        S   

         Last_name  
845         Abbing  
279         Abbott  
746         Abbott  
874        Abelson  
308        Abelson  
..             ...  
286      de Mulder  
282  de Pelsmaeker  
361      del Carlo  
153   van Billiard  
868  van Melkebeke  

[891 rows x 13 columns]

**4.A Do you find people with the same family name?**

Yes. Below we have also made a column called relatives, and sorted by which family name has the most members. Since this pulls the numbers registered about their family, it doesn't just add relatives based on the same name, which would obviously have the potential of giving false numbers.

In [11]:
df["Relatives"] = df["SibSp"]+df["Parch"]

In [12]:
df.sort_values("Relatives").tail(10)

PassengerId  Survived  Pclass                               Name     Sex  \
683          684         0       3        Goodwin, Mr. Charles Edward    male   
386          387         0       3    Goodwin, Master. Sidney Leonard    male   
480          481         0       3     Goodwin, Master. Harold Victor    male   
324          325         0       3           Sage, Mr. George John Jr    male   
863          864         0       3  Sage, Miss. Dorothy Edith "Dolly"  female   
201          202         0       3                Sage, Mr. Frederick    male   
159          160         0       3         Sage, Master. Thomas Henry    male   
792          793         0       3            Sage, Miss. Stella Anna  female   
180          181         0       3       Sage, Miss. Constance Gladys  female   
846          847         0       3           Sage, Mr. Douglas Bullen    male   

      Age  SibSp  Parch    Ticket   Fare Cabin Embarked Last_name  Relatives  
683  14.0      5      2   CA 2144  46.90   NaN        S   Goodwin          7  
386   1.0      5      2   CA 2144  46.90   NaN        S   Goodwin          7  
480   9.0      5      2   CA 2144  46.90   NaN        S   Goodwin          7  
324   NaN      8      2  CA. 2343  69.55   NaN        S      Sage         10  
863   NaN      8      2  CA. 2343  69.55   NaN        S      Sage         10  
201   NaN      8      2  CA. 2343  69.55   NaN        S      Sage         10  
159   NaN      8      2  CA. 2343  69.55   NaN        S      Sage         10  
792   NaN      8      2  CA. 2343  69.55   NaN        S      Sage         10  
180   NaN      8      2  CA. 2343  69.55   NaN        S      Sage         10  
846   NaN      8      2  CA. 2343  69.55   NaN        S      Sage         10

**5. USE PANDAS GROUPBY() OPERATIONS**

**5.A How many did travel on 1., 2. and 3rd class?**

In [13]:
df.groupby("Pclass")["PassengerId"].count()

Pclass
1    216
2    184
3    491
Name: PassengerId, dtype: int64

**5.B What travel class had the most deceased persons?**

In [14]:
df.groupby(["Pclass", "Survived"])["PassengerId"].count()

Pclass  Survived
1       0            80
        1           136
2       0            97
        1            87
3       0           372
        1           119
Name: PassengerId, dtype: int64

In first class had 80 dead, second class had 97 dead, and third class had 372 dead. So there were most dead in third class.  

**5.C How many women, men and children survived, listed by families?**

To do this, we will first make a feature called mfc_group (male, female, child). In this case we define child as someone of the age 15 and under, since that makes more sense for the 1910's. We will also make a data frame that only contains survivours, as that will make it easier to list. 

In [15]:
df["mfc_group"] = df["Sex"]
for x in df:
    df.loc[(df.Age<16), "mfc_group"]="child"

In [16]:
df_survivours = df[df["Survived"]==1]

In [17]:
df_survivours.groupby(["Last_name", "mfc_group"])["mfc_group"].count()

Last_name       mfc_group
Abbott          female       1
Abelson         female       1
Aks             female       1
Albimona        male         1
Allen           female       1
                            ..
Woolner         male         1
Yasbeck         child        1
Young           female       1
de Messemaeker  female       1
de Mulder       male         1
Name: mfc_group, Length: 314, dtype: int64